src/coap_openssl.c

```diff
@@ -1565,6 +1826,119 @@ tls_client_hello_call_back(SSL *ssl,
   }
   return SSL_CLIENT_HELLO_SUCCESS;
 }
+
+/*
+ * During the SSL/TLS initial negotiations, psk_tls_client_hello_call_back() is
+ * called early in the Client Hello processing so it is possible to determine
+ * whether SNI needs to be handled
+ *
+ * Set up by SSL_CTX_set_client_hello_cb().
+ */
+static int
+psk_tls_client_hello_call_back(SSL *ssl,
+                          int *al,
+                          void *arg UNUSED
+) {
+  coap_session_t *c_session;
+  coap_openssl_context_t *o_context;
+  coap_dtls_spsk_t *setup_data;
+  const unsigned char *out;
+  size_t outlen;
+
+  if (!ssl)
+    goto int_err;
+  c_session = (coap_session_t *)SSL_get_app_data(ssl);
+  if (!c_session || !c_session->context) {
+    goto int_err;
+  }
+  o_context = (coap_openssl_context_t *)c_session->context->dtls_context;
+  if (!o_context) {
+    goto int_err;
+  }
+  setup_data = &c_session->context->spsk_setup_data;
+
+  if (setup_data->validate_sni_call_back) {
+    /*
+     * SNI checking requested
+     */
+    const char *sni = "";
+    char *sni_tmp = NULL;
+    size_t i;
+    char lhint[COAP_DTLS_HINT_LENGTH];
+    coap_str_const_t lsni;
+
+    if (SSL_client_hello_get0_ext (ssl, TLSEXT_TYPE_server_name, &out, &outlen) &&
+        outlen > 5 &&
+        (((out[0]<<8) + out[1] +2) == (int)outlen) &&
+        out[2] == TLSEXT_NAMETYPE_host_name &&
+        (((out[3]<<8) + out[4] +2 +3) == (int)outlen)) {
+      /* Skip over length, type and length */
+      out += 5;
+      outlen -= 5;
+      sni_tmp = OPENSSL_malloc(outlen+1);
+      sni_tmp[outlen] = '\000';
+      memcpy(sni_tmp, out, outlen);
+      sni = sni_tmp;
+    }
+
+    lsni.length = strlen(sni);
+    lsni.s = (const uint8_t*)sni;
+    /* Is this a cached entry? */
+    for (i = 0; i < o_context->psk_sni_count; i++) {
+      if (coap_string_equal(&lsni, &o_context->psk_sni_entry_list[i].sni)) {
+        break;
+      }
+    }
+    if (i == o_context->psk_sni_count) {
+      /*
+       * New SNI request
+       */
+      const coap_dtls_spsk_key_t *new_entry = setup_data->validate_sni_call_back(
+                                               &lsni,
+                                               c_session,
+                                               setup_data->sni_call_back_arg);
+      if (!new_entry) {
+        *al = SSL_AD_UNRECOGNIZED_NAME;
+        return SSL_CLIENT_HELLO_ERROR;
+      }
+
+      o_context->psk_sni_entry_list =
+                OPENSSL_realloc(o_context->psk_sni_entry_list,
+                                (o_context->psk_sni_count+1)*sizeof(sni_entry));
+      o_context->psk_sni_entry_list[o_context->psk_sni_count].sni.length =
```

Comment:
What happens if `OPENSSL_realloc()` fails?
Also, ...sni.length must be cleared if the following `OPENSSL_strdup()` fails.

Reply:
Code now protected